[ENH] A new `ReconcilerForecaster` estimator for reconciling forecasts using base model residuals

ciaran-g opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/alan-turing-institute/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Added a new estimator `ReconcilerForecaster` which contains three reconciliation methods. See https://github.com/alan-turing-institute/sktime/issues/2157


#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->

A new `ReconcilerForecaster` class which is a composite of a base forecaster and a reconciliation method. The idea/methods are similar to the `Reconciler`, except here the reconciliation matrix is estimated using variations on the base-forecast residual covariance matrix. The three methods either use the sample covariance matrix `method="mint_cov"`, a shrunk covariance matrix `method="mint_shrink"`, or only the on-diagonal elements `method="wls_var"`.

#### Does your contribution introduce a new dependency? If yes, which one?

<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->

No

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

I was wondering if it would be easier or more intuitive for users if all of the reconciliation methods, including the transformers, were accessible though this estimator?

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] I've added unit tests and made sure they pass locally.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

<!--
Thanks for contributing!
-->

fkiraly commented on sktime/forecasting/reconcile.py:
```diff
@@ -0,0 +1,355 @@
+    "ciaran-g",
+]
+
+# todo: include the reconciler transformers?
```
that's not a silly idea at all!

You could import the transformer, and if a string for the "transformer-like" reconciliation is given, you could just do `Reconciler(strategy) * forecaster` inside. (i.e., I would not copy-paste code but wrap the transformer, optionally depending on strategy)

ciaran-g replied:
Ok cool 👍

fkiraly commented on sktime/forecasting/reconcile.py:
```diff
@@ -0,0 +1,355 @@
+
+    Reconciliation is applied to make the forecasts in a hierarchy of
+    time-series sum together appropriately.
+
```
maybe a few more lines how it relates to the `forecaster` arg, i.e., that forecasts are made by `forecaster` and then etc

ciaran-g replied:
added a few more lines

fkiraly commented on sktime/forecasting/reconcile.py:
```diff
@@ -0,0 +1,355 @@
+        """
+        self._check_method()
+
+        # # check the length of index if not hierarchical just return forecaster
```
well, not 100% precise, you fit the forecaster and return reference to self (which is the correct thing to do)

fkiraly commented on sktime/forecasting/reconcile.py:
```diff
@@ -0,0 +1,355 @@
+
+from sktime.forecasting.base import BaseForecaster, ForecastingHorizon
+from sktime.transformations.hierarchical.aggregate import _check_index_no_total
+from sktime.transformations.hierarchical.reconcile import (
```
hm... should this move somewhere? The "parent child", or more generally, graph stuff, could fit in `datatypes` as utilities for the hierarchical data. Although, for the moment, let´s leave it where it is to avoid refactor complexity sneaking in. Worth thinking about though.

fkiraly commented on sktime/forecasting/reconcile.py:
```diff
@@ -0,0 +1,355 @@
+        # parent child df
+        self.parent_child = _parent_child_df(self.s_matrix)
+
+        # bug in self.forecaster_.predict_residuals() for heir data
```
oops - well, can we fix it there?

fkiraly commented on sktime/forecasting/reconcile.py:
```diff
@@ -0,0 +1,355 @@
+        )
+        self.residuals = y - self.forecaster_.predict(fh=fh_resid, X=X)
+
+        if self.method == "mint_cov":
```
instead of having strings, would it make sense to have `shrink` and `diag_only` directly as hyperparameters?
Although the current solution seems better to me if we will have methods here that do not use `_get_g_matrix_mint` or a simple extension, in the future.

ciaran-g replied:
Yeah I think we should maybe leave it as is then? I agree it will be easier to add methods that use the residuals or training data but not the covariance matrix, e.g. top down historical proportions.

fkiraly replied:
ok, makes sense

fkiraly commented on sktime/forecasting/reconcile.py:
```diff
@@ -0,0 +1,355 @@
+        fh_resid = ForecastingHorizon(
+            y.index.get_level_values(-1).unique(), is_relative=False
+        )
+        self.residuals = y - self.forecaster_.predict(fh=fh_resid, X=X)
```
this should be `residuals_` according to the condition that this is a fitted parameter of the model

fkiraly commented on sktime/forecasting/reconcile.py:
```diff
@@ -0,0 +1,355 @@
+        -------
+        self : reference to self
+        """
+        if y.index.nlevels < 2:
```
the update seems incorrect or problematic.

Let's say we update only with a single time point, i.e., a length 1 `y`, then we have only a single residuals and compute matrices on only a single residuals!

Should we not use, say, `self._y` instead? In which case this would be equivalent to `fit` on all seen data, but the matrix updates could be done more efficiently, e.g., by the WSM formula. (although that might be overkill for an initial PR?)

fkiraly commented on sktime/forecasting/reconcile.py:
```diff
@@ -0,0 +1,355 @@
+        # cov matrix
+        resid = resid.unstack().transpose()
+        nobs = len(resid)
+        cov_mat = resid.transpose().dot(resid) / (nobs - 1)
```
can't we use numpy.cov here?

ciaran-g replied:
yep, I simplified this section a wee bit! 👍

fkiraly reviewed (changes requested):
Looks great, thanks!

Blocking issue: I think `update` is not correct?

Other things are not blocking.

Further, can you kindly point me to the precise location fo the shrinkage formulae in the references? The things you are implementing starting line 272.

ciaran-g commented:
Just updated @fkiraly with changes to the update function, added the `Reconciler()` transformers, and simplified the shrinkage calc a wee bit.

The shrinkage calculation is detailed in Appendix A of this [paper](https://doi.org/10.2202/1544-6115.1175). I was originally doing it the calculation the long way round as detailed in the paper when I started off, but realised you can do it faster with some re-arranging thanks to the folks at the [fable](https://github.com/tidyverts/fabletools) R package

ciaran-g commented:
Yep, just updated to also store the matrices in `ReconcilerForecaster()` if the reconciliation method is one of the transformers, so it matches the other methods

fkiraly reviewed (approved):
`update` looks good now, thanks.

Non-blocking comment:

* what `self.forecaster` is is not consistent across the methods. One time, it is the fitted forecaster, one time it is the entire pipeline. I would ensure it is the pipeline in both cases, i.e., either bind it to the fittable pipeline element, or execute the pipeline manually.
* The `self.forecaster_` attribute should be mentioned in the docstring.

fkiraly commented:
moved the remaining points here: https://github.com/alan-turing-institute/sktime/issues/2875 and merged, so it goes in for 0.12.1

aiwalter commented:
@ciaran-g do you mind to add this new forecaster also to docs?
